### PR TITLE
Add the support for kafka in cAdvisor's storage, including output of container id and labels

### PR DESCRIPTION
**Pull request by arichardet**
Steps to run and configure with custom broker and topic. Default broker is `localhost:9092`. Default topic is `stats`.
`sudo ./cadvisor -storage_driver=kafka -storage_driver_kafka_broker_list=localhost:9092 -storage_driver_kafka_topic=stats`

This will push json encoded container statistics onto the designated topic. Includes additional output of container id and labels, added to the v1 container spec.

### CODE REVIEW

**Comment by arichardet** [comment]
I signed it!

File: info/v1/container.go [hunk 1]
```diff
@@ -64,10 +64,16 @@ type ContainerSpec struct {
...
 	Image string `json:"image,omitempty"`
+
+	// The ID of the container
+	ContainerId string `json:"container_id,omitempty"`
 }
 
 // Container reference contains enough information to uniquely identify a container
 type ContainerReference struct {
+	// The container id
+	Id string `json:"id"`
```
Comment (timstclair) [inline info/v1/container.go]: Why is this required for a reference, but optional for the spec? I think these should be consistent, or at least the Reference should have weaker requirements than the spec.

File: storage/common_flags.go [hunk 2]
```diff
@@ -26,3 +26,5 @@ var ArgDbName = flag.String("storage_driver_db", "cadvisor", "database name")
 var ArgDbTable = flag.String("storage_driver_table", "stats", "table name")
 var ArgDbIsSecure = flag.Bool("storage_driver_secure", false, "use secure connection with database")
 var ArgDbBufferDuration = flag.Duration("storage_driver_buffer_duration", 60*time.Second, "Writes in the storage driver will be buffered for this duration, and committed to the non memory backends as a single transaction")
+var ArgKafkaBrokerList = flag.String("storage_driver_kafka_broker_list", "localhost:9092", "kafka broker(s) csv")
+var ArgKafkaTopic = flag.String("storage_driver_kafka_topic", "stats", "kafka topic")
```
Comment (timstclair) [inline storage/common_flags.go]: I don't think these are common flags. They should be in the kafka package (e.g. like [bigquery](https://github.com/google/cadvisor/blob/master/storage/bigquery/client/client.go#L28))

File: storage/kafka/kafka.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
```
Comment (timstclair) [inline storage/kafka/kafka.go]: 2016 :)

File: storage/kafka/kafka.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	MachineName     string               `json:"machine_name,omitempty"`
+	ContainerName   string               `json:"container_Name,omitempty"`
+	ContainerId     string               `json:"container_Id,omitempty"`
+	ContainerLabels map[string]string    `json:"container_labels,omitempty"`
+	ContainerStats  *info.ContainerStats `json:"container_stats,omitempty"`
+}
+
+func (driver *kafkaStorage) containerStatsAndDefaultValues(ref info.ContainerReference, stats *info.ContainerStats) *detailSpec {
+	timestamp := stats.Timestamp.UnixNano() / 1E3
+	var containerName string
```
Comment (timstclair) [inline storage/kafka/kafka.go]: (nit) move this down to where it's used (just before the `if`)

File: storage/kafka/kafka.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+type detailSpec struct {
+	Timestamp       int64                `json:"timestamp"`
+	MachineName     string               `json:"machine_name,omitempty"`
+	ContainerName   string               `json:"container_Name,omitempty"`
+	ContainerId     string               `json:"container_Id,omitempty"`
+	ContainerLabels map[string]string    `json:"container_labels,omitempty"`
+	ContainerStats  *info.ContainerStats `json:"container_stats,omitempty"`
+}
+
+func (driver *kafkaStorage) containerStatsAndDefaultValues(ref info.ContainerReference, stats *info.ContainerStats) *detailSpec {
```
Comment (timstclair) [inline storage/kafka/kafka.go]: This method name confusing to me. Consider changing it to something like `createDetailSpec`, `getDetailSpec`, or `infoToDetailSpec`.

File: storage/kafka/kafka.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	storage.RegisterStorageDriver("kafka", new)
+}
+
+type kafkaStorage struct {
+	producer    kafka.SyncProducer
+	topic       string
+	machineName string
+}
+
+type detailSpec struct {
```
Comment (timstclair) [inline storage/kafka/kafka.go]: This is basically the exported API, right? Would it be useful to put this in its own package & export it, so that go kafka consumers can import the type?

File: storage/kafka/kafka.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	Timestamp       int64                `json:"timestamp"`
```
Comment (timstclair) [inline storage/kafka/kafka.go]: Document the format of the timestamp. Can you use time.Time instead?

File: storage/kafka/kafka.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	MachineName     string               `json:"machine_name,omitempty"`
+	ContainerName   string               `json:"container_Name,omitempty"`
+	ContainerId     string               `json:"container_Id,omitempty"`
+	ContainerLabels map[string]string    `json:"container_labels,omitempty"`
+	ContainerStats  *info.ContainerStats `json:"container_stats,omitempty"`
+}
+
+func (driver *kafkaStorage) containerStatsAndDefaultValues(ref info.ContainerReference, stats *info.ContainerStats) *detailSpec {
+	timestamp := stats.Timestamp.UnixNano() / 1E3
```
Comment (timstclair) [inline storage/kafka/kafka.go]: Microseconds is a bit of an unusual unit - is that a Kafka thing?

Also, for clarity, prefer:
```
timestamp := stats.Timestamp.UnixNano() / time.Microsecond
```

File: storage/kafka/kafka.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+	var containerName string
+	containerId := ref.Id
+	containerLabels := ref.Labels
+
+	if len(ref.Aliases) > 0 {
+		containerName = ref.Aliases[0]
+	} else {
+		containerName = ref.Name
+	}
```
Comment (timstclair) [inline storage/kafka/kafka.go]: Can you move this block to a shared utility function? Something like `GetPreferredName`. It's non-obvious to me that in practice `Name` is usually the ID, and `Aliases[0]` is a more accurate representation of the container "name".

File: storage/kafka/kafka.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+		ContainerId:     containerId,
+		ContainerLabels: containerLabels,
+		ContainerStats:  stats,
+	}
+	return detail
+}
+
+func (driver *kafkaStorage) AddStats(ref info.ContainerReference, stats *info.ContainerStats) error {
+	detail := driver.containerStatsAndDefaultValues(ref, stats)
+	b, _ := json.Marshal(detail)
```
Comment (timstclair) [inline storage/kafka/kafka.go]: Is there a reason to ignore the error here? I think at the very least you should log it.

File: storage/kafka/kafka.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+		Topic: driver.topic,
+		Value: kafka.StringEncoder(b),
+	}
+
+	_, _, err := driver.producer.SendMessage(msg)
+	if err != nil {
+		fmt.Println("kafka storage error:", err)
+	}
+
+	if stats == nil {
```
Comment (timstclair) [inline storage/kafka/kafka.go]: This if is unnecessary, as you return `nil` below.

File: storage/kafka/kafka.go [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+func (driver *kafkaStorage) AddStats(ref info.ContainerReference, stats *info.ContainerStats) error {
+	detail := driver.containerStatsAndDefaultValues(ref, stats)
+	b, _ := json.Marshal(detail)
+	msg := &kafka.ProducerMessage{
+		Topic: driver.topic,
+		Value: kafka.StringEncoder(b),
+	}
+
+	_, _, err := driver.producer.SendMessage(msg)
```
Comment (timstclair) [inline storage/kafka/kafka.go]: `AddStats(..)` should not block. Can you just use the `kafka.AsyncProducer` instead?

File: storage/kafka/kafka.go [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	return nil
+}
+
+func (self *kafkaStorage) Close() error {
+	return self.producer.Close()
+}
+
+func new() (storage.StorageDriver, error) {
```
Comment (timstclair) [inline storage/kafka/kafka.go]: I'm surprised that go allows this, since `new` is a built-in. Since I'm confused as to what function `new` actually refers to in this file, I'd prefer to call this something like `newKafka` or `newDriver`

Comment (timstclair) [inline storage/kafka/kafka.go]: or `newStorage` :) You can just inline newStorage in this method.

File: storage/kafka/kafka.go [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+func newStorage(machineName,
+	brokers,
+	topic string,
+) (storage.StorageDriver, error) {
+	config := kafka.NewConfig()
+	config.Producer.RequiredAcks = kafka.WaitForAll
+	config.Producer.Retry.Max = 10
+
+	brokerList := strings.Split(brokers, ",")
+	fmt.Println("Kafka brokers:", strings.Join(brokerList, ", "))
```
Comment (timstclair) [inline storage/kafka/kafka.go]: This should be a log statement (probably info level 4).

Comment (timstclair) [inline storage/kafka/kafka.go]: Just log brokers, rather than joining on the same character you split on.

**Comment by carmark** [comment]
could you please re-order the import packages as the follow order:
* golang self package
* cadvisor package
* third-party package

**Comment by timstclair** [comment]
Did you run `godep save`? I think you're missing the sarama deps.

**Comment by timstclair** [comment]
@k8s-bot ok to test

File: storage/kafka/kafka.go (hunk 13 above)
Comment (arichardet) [inline storage/kafka/kafka.go]: I was aiming to be consistent with the convention set by [this refactor](https://github.com/google/cadvisor/commit/b0ce5720b03b009c4bc48a57857c0737cb2ae07f).

Comment (timstclair) [inline storage/kafka/kafka.go]: I see, that should probably be fixed. But, consistency trumps, so I'm ok with what you have for now. Let me know when you're ready for me to take another look.

**Comment by arichardet** [comment]
@timstclair I've addressed your PR feedback.

**Comment by arichardet** [comment]
Travis-CI is not liking the Kafka.go header going from 2015 -> 2016

**Comment by timstclair** [comment]
Oh yeah, that needs to be fixed.

**Comment by timstclair** [comment]
It looks like you somehow picked up a bunch of extraneous commits. Can you try rebasing?

**Comment by arichardet** [comment]
@timstclair I tried rebasing and it this is what I got :-1: I can close this PR and resubmit as a fresh one that is cleanly in one commit if thats ok.

**Comment by timstclair** [comment]
It looks like your commits for this PR are interspersed with merged commits, in which case you might need to manually repair it. I'd suggest trying `git rebase -i HEAD~35` and moving all your commits so they are grouped together at the bottom. If that doesn't fix it, try rebasing again using the instructions here: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md#keeping-your-development-fork-in-sync

**Comment by arichardet** [comment]
@timstclair I was able to clean it up.

**Comment by timstclair** [comment]
Great, but unfortunately github won't show the diff now because of all the godeps changes. Could you please move the godeps changes to a separate commit, so that I can review the other files separately? Thanks!

**Comment by arichardet** [comment]
@timstclair I've separated them.

**Comment by arichardet** [comment]
@timstclair Is there anything else you need me to do?

**Comment by arichardet** [comment]
@timstclair should close this and resubmit a clean PR? Thanks

**Comment by timstclair** [comment]
Sorry, I was out of office last week. I'll take a final look at this later today.

**Comment by timstclair** [comment]
Just a few more comments. Thanks for your patience.

**Comment by arichardet** [comment]
@timstclair comments addressed. Thanks

**Comment by timstclair** [comment]
LGTM, just 1 small comment

**Comment by arichardet** [comment]
@timstclair comment addressed. Thanks

**Comment by timstclair** [comment]
LGTM. Thanks!

**Comment by timstclair** [comment]
@k8s-bot ok to test
